### ISSUE TITLE
Easier Extension Adding

### ISSUE THREAD

**Issue by titusfortner** [mozilla/geckodriver#1476]
This is a feature request, feel free to close / redirect me if there's a better place to do this.

For reference, the Selenium Ruby code no longer seems to work for using `Firefox::Profile#add_extension`. I see that there's an endpoint to do this now, but this really seems like something that belongs in the capabilities information, not changed after opening the browser.

Ideally we wouldn't need "complicated" code to create a whole profile, we should be able to just encode the extension and pass it along in the capabilities like this:

```
{
    "capabilities": {
        "alwaysMatch": {
            "moz:firefoxOptions": {
                "extensions": {
                    [base_64_string_extension_one, base_64_string_extension_two]
                }
            }
        }
    }
}
```

Alternatively, we (I) need to figure out how to fix the Ruby code. :)

**Comment by andreastt** [comment]
I don’t think I’m opposed to this idea, but it would be a fairly low priority since we already have `POST /session/{session id}/addon/install`.

**Comment by whimboo** [comment]
I feel this only adds confusion so I agree with Andreas. AFAIK also all other bindings make use of that specific end-point Andreas pointed out, and we haven't heard complains about it not working.

**Comment by titusfortner** [comment]
@whimboo that's incorrect.
As far as I know, Chrome is the only other browser that supports adding extensions, and it is definitely done in the capabilities as a path to the extension: http://chromedriver.chromium.org/extensions

As a tester I feel very strongly that everything related to the browser setup should be able to go into a config file and loaded when initializing the session. The solution you provide will force me to add conditionals to my test logic and likely hard code the values I need somewhere independent of the rest of my configuration code.

For example, I'm putting everything I need [in a yaml](https://github.com/saucelabs-sample-test-frameworks/Ruby-RSpec-Watir/blob/master/spec/platforms.yml), then [pull values from it](https://github.com/saucelabs-sample-test-frameworks/Ruby-RSpec-Watir/blob/master/spec/spec_helper.rb#L5-L11), then [merge it in](https://github.com/saucelabs-sample-test-frameworks/Ruby-RSpec-Watir/blob/master/spec/spec_helper.rb#L36). This means favoring lookups over conditionals, reducing the complexity of the initialization code.

**Comment by whimboo** [comment]
I see. That would indeed be helpful then. Sorry, if I misunderstood it before.

In that case we can offer that `extension` property, and let geckodriver install the listed extensions right after creating the session by using the internal `Addon:Install` command.

**Comment by jps042** [comment]
Was this issue ever resolved? I am able to specify adding an extension in chrome through chromeOptions like 
Selenium::WebDriver::Remote::Capabilities.chrome({
                                                                               'chromeOptions' => {
                                                                                   'extensions' => [
                                                                                       Base64.strict_encode64(File.open(crx_path,'rb').read)
                                                                                   ],
                                                                                   prefs: prefs,
                                                                                   args: switches,
                                                                                   detach: true
                                                                               }
                                                                           })

and I would like to be able to do the same for firefoxOptions like:
Selenium::WebDriver::Remote::Capabilities.firefox({
                                                                                   "capabilities": {
                                                                                       "alwaysMatch": {
                                                                                           "moz:firefoxOptions": {
                                                                                               'extension' => [
                                                                                                   Base64.strict_encode64(File.open(xpi_path,'rb').read)
                                                                                               ],
                                                                                               args: switches,
                                                                                               prefs: prefs,
                                                                                               log: {
                                                                                                   "level": "trace"
                                                                                               }
                                                                                           }
                                                                                       }
                                                                                   }
                                                                               })
Any update would be much appreciated :)

**Comment by whimboo** [comment]
No, it's not fixed but I filed https://bugzilla.mozilla.org/show_bug.cgi?id=1561841 to track this request internally.

**Comment by pandasm** [comment]
This one is open from an year, can we get any progress or any workaround ?

**Comment by whimboo** [comment]
We are currently not working on that due to other priorities. As workaround you can still use the vendor endpoint `/session/{sessionId}/moz/addon/install` to install an add-on.

**Comment by bonigarcia** [comment]
+1 to this feature. See this [issue](10946) for some test examples that would benefit with the installation of web extensions with firefox options.

**Comment by titusfortner** [comment]
Wrong link in @bonigarcia's comment: https://github.com/seleniumhq/selenium/issues/10946

**Comment by whimboo** [comment]
@titusfortner and @christian-bromann is https://bugzilla.mozilla.org/show_bug.cgi?id=1561841#c4 what you would like to see? Any other feedback from your side? Thanks.

**Comment by whimboo** [comment]
@nechaev-chromium given that Maxim isn't around right now maybe you could have a look at https://bugzilla.mozilla.org/show_bug.cgi?id=1561841#c4 and let me know if that is something you might consider as well for chromedriver in case Chrome requires such additional permissions? Thanks.

**Comment by christian-bromann** [comment]
> @christian-bromann is https://bugzilla.mozilla.org/show_bug.cgi?id=1561841#c4 what you would like to see? Any other feedback from your side? Thanks.

What you have [proposed](https://bugzilla.mozilla.org/show_bug.cgi?id=1561841#c4) sounds good to me. I think there is value having the ability to install a web extension on session start as well as during runtime as long as we find a common specification and browser handling for it.

**Comment by titusfortner** [comment]
Previous code bases I've used, the configuration files and the browser specific conditionals were all applied and managed before the driver started. Obviously it can be worked around, but I've heard more complaints that Firefox can't install extensions in the capabilities than that chrome can't uninstall extensions.

**Comment by whimboo** [comment]
Thank you all for the feedback. Looks like we should find a common way to install webextensions via capabilities.

> @nechaev-chromium given that Maxim isn't around right now maybe you could have a look at https://bugzilla.mozilla.org/show_bug.cgi?id=1561841#c4 and let me know if that is something you might consider as well for chromedriver in case Chrome requires such additional permissions? Thanks.

@mathiasbynens maybe you could have a look? Not sure whom else to ask around chromedriver's implementation while Maksim is out. Thanks.

**Comment by whimboo** [comment]
@sadym-chromium mind having a look at my last comment? Thanks!

**Comment by sadym-chromium** [comment]
@whimboo next week, after my vacation. Or feel free to ping other our folks

**Comment by sadym-chromium** [comment]
I'm not sure if it's possible to implement `"extensionsInstallTemporarily": false` in Chrome. WRT `extensionsAllowPrivateBrowsing`, is it still required, if we have UserContexts?

**Comment by whimboo** [comment]
Maybe lets follow-up on https://github.com/w3c/webdriver-bidi/issues/548 for further discussion given that this is basically a spec and not a driver issue.
